tests/test_ensemble.py

```diff
@@ -0,0 +1,381 @@
+        g_aa = [3,7,4]
+        g_bb = [1,9,3]
+        ens.rdf['A','B'] = relentless.RDF(r=r, g=g_ab)
+        ens.rdf['A','A'] = relentless.RDF(r=r, g=g_aa)
+        ens.rdf['B','B'] = relentless.RDF(r=r, g=g_bb)
+        ens.clear()
+        self.assertDictEqual(ens.rdf['A','B'], {})
+        self.assertDictEqual(ens.rdf['A','A'], {})
+        self.assertDictEqual(ens.rdf['B','B'], {})
+
+    def test_copy(self):
+        """Test copy method"""
+        v_obj = relentless.Cube(L=1.0)
+
+        #P and mu set
+        ens = relentless.Ensemble(T=10, P=2.0, mu={'A':0.1,'B':0.2})
+        ens.V = v_obj
+        ens.N['A'] = 1
+        ens.N['B'] = 2
+        ens_ = ens.copy()
+        assert ens_ is not ens
+        self.assertCountEqual(ens.types, ens_.types)
+        self.assertCountEqual(ens.rdf.pairs, ens_.rdf.pairs)
+        self.assertAlmostEqual(ens.T, ens_.T)
+        self.assertAlmostEqual(ens.P, ens_.P)
+        self.assertIsInstance(ens_.V, relentless.Cube)
+        self.assertAlmostEqual(ens.V.volume, ens_.V.volume)
+        self.assertDictEqual(ens.mu.todict(), ens_.mu.todict())
+        self.assertDictEqual(ens_.N.todict(), ens.N.todict())
+
+        #V and N set
+        ens = relentless.Ensemble(T=20, V=v_obj, N={'A':1,'B':2})
+        ens.P = 1.0
+        ens.mu['A'] = 0.1
+        ens.mu['B'] = 0.2
+        ens_ = ens.copy()
+        assert ens_ is not ens
+        self.assertCountEqual(ens.types, ens_.types)
+        self.assertCountEqual(ens.rdf.pairs, ens_.rdf.pairs)
+        self.assertAlmostEqual(ens.T, ens_.T)
+        self.assertAlmostEqual(ens.P, ens_.P)
+        self.assertIsInstance(ens_.V, relentless.Cube)
+        self.assertAlmostEqual(ens.V.volume, ens_.V.volume)
+        self.assertDictEqual(ens.mu.todict(), ens_.mu.todict())
+        self.assertDictEqual(ens_.N.todict(), ens.N.todict())
+
+        #mu and N used for one type each
+        ens = relentless.Ensemble(T=100, V=v_obj, mu={'A':0.1}, N={'B':2})
+        ens.P = 1.0
+        ens.mu['B'] = 0.2
+        ens.N['A'] = 1
+        ens_ = ens.copy()
+        assert ens_ is not ens
+        self.assertCountEqual(ens.types, ens_.types)
+        self.assertCountEqual(ens.rdf.pairs, ens_.rdf.pairs)
+        self.assertAlmostEqual(ens.T, ens_.T)
+        self.assertAlmostEqual(ens.P, ens_.P)
+        self.assertIsInstance(ens_.V, relentless.Cube)
+        self.assertAlmostEqual(ens.V.volume, ens_.V.volume)
+        self.assertDictEqual(ens.mu.todict(), ens_.mu.todict())
+        self.assertDictEqual(ens_.N.todict(), ens.N.todict())
+
+        #test copying rdf
+        ens = relentless.Ensemble(T=100, V=v_obj, mu={'A':0.1}, N={'B':2})
+        r = [1,2,3]
+        g_ab = [2,9,5]
+        g_aa = [3,7,4]
+        g_bb = [1,9,3]
+        ens.rdf['A','B'] = relentless.RDF(r=r, g=g_ab)
+        ens.rdf['A','A'] = relentless.RDF(r=r, g=g_aa)
+        ens.rdf['B','B'] = relentless.RDF(r=r, g=g_bb)
+        ens_ = ens.copy()
+        assert ens_ is not ens
```

Comment:
You should additionally check that `ens.rdf is not ens_.rdf` to make sure the deep copy worked. (A shallow copy should fail this check.)